gcloud/storage/batch.py

```diff
@@ -146,14 +216,49 @@ def finish(self):
 
         # Strip off redundant header text
         _, body = payload.split('\n\n', 1)
-        headers = dict(multi._headers)
+        return dict(multi._headers), body
+
+    def _finish_futures(self, responses):
+        """Apply all the batch responses to the futures created.
+
+        :type responses: list of (headers, payload) tuples.
+        :param responses: List of headers and payloads from each response in
+                          the batch.
+
+        :raises: :class:`ValueError` if no requests have been deferred.
+        """
+        # If a bad status occurs, we track it, but don't raise an exception
+        # until all futures have been populated.
+        exception_args = None
+
+        if len(self._futures) != len(responses):
+            raise ValueError('Expected a response for every request.')
+
+        for future, sub_response in zip(self._futures, responses):
+            resp_headers, sub_payload = sub_response
+            if not 200 <= resp_headers.status < 300:
+                exception_args = exception_args or (resp_headers,
+                                                    sub_payload)
+            future.set_future_value(sub_payload)
```

Comment:
@tseaver comment from f2a9c7dfa5f753fc007a8cea4741aa99131da5bb:

> ISTM that we should not be assigning values for exception returns.

----

I agree. It's as simple as putting this statement in an `else:`, yes?